src/OrbitGl/OrbitApp.cpp

```diff
@@ -2291,35 +2291,54 @@ void OrbitApp::UpdateAfterSymbolLoading() {
     return;
   }
-  const CaptureData& capture_data = GetCaptureData();
+  Future<std::shared_ptr<PostProcessedSamplingData>> post_processed_sampling_data_future =
+      thread_pool_->Schedule([this]() {
+        const CaptureData& capture_data = GetCaptureData();
 
-  PostProcessedSamplingData post_processed_sampling_data =
-      orbit_client_model::CreatePostProcessedSamplingData(capture_data.GetCallstackData(),
-                                                          capture_data, *module_manager_);
-  GetMutableCaptureData().set_post_processed_sampling_data(post_processed_sampling_data);
-  auto selection = std::make_unique<SelectionData>(*module_manager_, GetCaptureData(),
-                                                   GetCaptureData().post_processed_sampling_data(),
-                                                   &GetCaptureData().GetCallstackData());
-  main_window_->SetTopDownView(selection->GetTopDownView());
-  main_window_->SetBottomUpView(selection->GetBottomUpView());
-  main_window_->UpdateSamplingReport(&selection->GetCallstackData(),
-                                     &selection->GetPostProcessedSamplingData());
-  full_capture_selection_ = std::move(selection);
-
-  main_window_->ClearCallstackInspection();
-  inspection_selection_.reset();
-  time_range_thread_selection_.reset();
-
-  PostProcessedSamplingData selection_post_processed_sampling_data =
-      orbit_client_model::CreatePostProcessedSamplingData(capture_data.selection_callstack_data(),
-                                                          capture_data, *module_manager_);
-  GetMutableCaptureData().set_selection_post_processed_sampling_data(
-      std::move(selection_post_processed_sampling_data));
-
-  SetSelectionTopDownView(capture_data.selection_post_processed_sampling_data(), capture_data);
-  SetSelectionBottomUpView(capture_data.selection_post_processed_sampling_data(), capture_data);
-  main_window_->UpdateSelectionReport(&capture_data.selection_callstack_data(),
-                                      &capture_data.selection_post_processed_sampling_data());
+        return std::make_shared<PostProcessedSamplingData>(
+            orbit_client_model::CreatePostProcessedSamplingData(capture_data.GetCallstackData(),
+                                                                capture_data, *module_manager_));
+      });
+  Future<std::shared_ptr<std::unique_ptr<SelectionData>>> selection_future =
+      post_processed_sampling_data_future.Then(
+          thread_pool_.get(),
+          [this](const std::shared_ptr<PostProcessedSamplingData>& post_processed_sampling_data) {
+            GetMutableCaptureData().set_post_processed_sampling_data(
+                std::move(*post_processed_sampling_data));
+            auto selection = std::make_unique<SelectionData>(
+                *module_manager_, GetCaptureData(), GetCaptureData().post_processed_sampling_data(),
+                &GetCaptureData().GetCallstackData());
```

Comment:
Can we be sure that nothing mutates the capture data in the mean time?

A quick search for `GetMutableCaptureData` tells me that there a lot of places that mutate CaptureData. Not sure if we can just access it from a different thread and rely on thread compatibility.

Reply:
I had a closer look. It seems to me the obvious step would be to make `CaptureData` thread safe. But that's rather hard. Many getters return a `const T&` and I am not sure if we can afford return by value here - on top of the lock overhead.

Another option to explore might be to remove the dependency from `CaptureData`. This might require either moving some parts into `PostProcessedSamplingData`, or replicating into `PostProcessedSamplingData` and refactor into new types. Of course this also sounds like a lot of work.